### PR TITLE
Fix 5s delay on sticky queue

### PR DESCRIPTION
**Pull request by yiminc**
**What changed?**
Fix the 5s delay after sticky worker is gone

**Why?**
Sticky worker mark workflows as running on sticky queue, but when sticky worker is gone, they don't notify server. When new workflow task or query is scheduled, server still dispatch them to the old sticky queue which now has no worker working on and will only result in 5s sticky queue's schedule_to_start timeout. No mater how long after the old sticky worker is gone, there is always that 5s delay. 
This PR fix that by checking if if there is no inflight poller for the sticky queue, and no poller is seem within 10s. In that case, matching would return a new StickyWorkerUnavailable error and history would retry that by using original task queue.

**How did you test it?**
New integration test to cover both workflow case and query case.

**Potential risks**
No

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/matching/matchingEngine.go [hunk 1]
```diff
@@ -259,6 +255,13 @@ func (e *matchingEngineImpl) AddWorkflowTask(
 		return false, err
 	}
 
+	if taskQueueKind == enumspb.TASK_QUEUE_KIND_STICKY {
+		// check if sticky worker is gone, if so fail this request, caller should use original task queue
+		if !tlMgr.HasRecentPoller(time.Now().Add(-stickyPollerUnavailableWindow)) {
```
Comment (yux0) [inline service/matching/matchingEngine.go:260]: nit .Sub(stickyPollerUnavailableWindow)

File: service/matching/matchingEngine.go [hunk 2]
```diff
@@ -542,6 +545,13 @@ func (e *matchingEngineImpl) QueryWorkflow(
 	if err != nil {
 		return nil, err
 	}
+	if taskQueueKind == enumspb.TASK_QUEUE_KIND_STICKY {
```
Comment (yux0) [inline service/matching/matchingEngine.go:548]: what about AddAcitiviyTask?

File: service/matching/taskQueueManager.go [hunk 3]
```diff
@@ -96,6 +96,7 @@ type (
 		DispatchQueryTask(ctx context.Context, taskID string, request *matchingservice.QueryWorkflowRequest) (*matchingservice.QueryWorkflowResponse, error)
 		CancelPoller(pollerID string)
 		GetAllPollerInfo() []*taskqueuepb.PollerInfo
+		HasRecentPoller(earliestAccessTime time.Time) bool
```
[the remark this answers is not in the document]
Comment (yiminc) [reply service/matching/taskQueueManager.go:99]: Should be HasPollerAfter :)

File: service/matching/matchingEngine.go (hunk 2 above)
Comment (yiminc) [reply service/matching/matchingEngine.go:548]: Activity never use sticky queues.

File: service/matching/matchingEngine.go (hunk 1 above)
Comment (yiminc) [reply service/matching/matchingEngine.go:260]: for t.Sub(u), u is time not duration.

File: service/history/transferQueueActiveTaskExecutor.go [hunk 4]
```diff
@@ -235,7 +236,21 @@ func (t *transferQueueActiveTaskExecutor) processWorkflowTask(
...
+
+	if _, ok := err.(*serviceerrors.StickyWorkerUnavailable); ok {
+		// sticky worker is unavailable, switch to original task queue
+		taskQueue = &taskqueuepb.TaskQueue{
+			// do not use task.TaskQueue which is sticky, use original task queue from mutable state
+			Name: mutableState.GetExecutionInfo().TaskQueue,
+			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+		}
+		workflowRunTimeout := timestamp.DurationValue(executionInfo.WorkflowRunTimeout)
+		taskScheduleToStartTimeoutSeconds = int64(workflowRunTimeout.Round(time.Second).Seconds())
```
Comment (yycptt) [inline service/history/transferQueueActiveTaskExecutor.go:250]: The schedule to start timer is still there. Do we want to use that as the ttl?

File: service/history/transferQueueActiveTaskExecutor.go [hunk 5]
```diff
@@ -235,7 +236,21 @@ func (t *transferQueueActiveTaskExecutor) processWorkflowTask(
...
 	release(nil)
-	return t.pushWorkflowTask(ctx, task, taskQueue, timestamp.DurationFromSeconds(taskScheduleToStartTimeoutSeconds))
+
+	err = t.pushWorkflowTask(ctx, task, taskQueue, timestamp.DurationFromSeconds(taskScheduleToStartTimeoutSeconds))
+
+	if _, ok := err.(*serviceerrors.StickyWorkerUnavailable); ok {
+		// sticky worker is unavailable, switch to original task queue
+		taskQueue = &taskqueuepb.TaskQueue{
+			// do not use task.TaskQueue which is sticky, use original task queue from mutable state
+			Name: mutableState.GetExecutionInfo().TaskQueue,
```
Comment (yycptt) [inline service/history/transferQueueActiveTaskExecutor.go:246]: I understand currently this field and the WorkflowRunTimeout below are read only fields. But I don't think it's a good practice to access mutable state after releasing the workflow lock.

File: service/history/transferQueueActiveTaskExecutor.go [hunk 6]
```diff
@@ -235,7 +236,21 @@ func (t *transferQueueActiveTaskExecutor) processWorkflowTask(
...
+			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+		}
+		workflowRunTimeout := timestamp.DurationValue(executionInfo.WorkflowRunTimeout)
+		taskScheduleToStartTimeoutSeconds = int64(workflowRunTimeout.Round(time.Second).Seconds())
+		err = t.pushWorkflowTask(ctx, task, taskQueue, timestamp.DurationFromSeconds(taskScheduleToStartTimeoutSeconds))
```
Comment (yycptt) [inline service/history/transferQueueActiveTaskExecutor.go:251]: nit: it took me sometime to realize why calling reset stickyness api is unnecessary here. maybe add some comments explaining the reasoning?

File: service/history/transferQueueActiveTaskExecutor.go (hunk 5 above)
Comment (yiminc) [reply service/history/transferQueueActiveTaskExecutor.go]: good point. Will move it to before lock release.

File: service/history/transferQueueActiveTaskExecutor.go (hunk 6 above)
Comment (yiminc) [reply service/history/transferQueueActiveTaskExecutor.go:256]: added comments

File: service/history/transferQueueStandbyTaskExecutor.go [hunk 7]
```diff
@@ -166,24 +166,12 @@ func (t *transferQueueStandbyTaskExecutor) processWorkflowTask(
...
-				Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
-			}
-			workflowRunTimeout := timestamp.DurationValue(executionInfo.WorkflowRunTimeout)
-			taskScheduleToStartTimeoutSeconds = int64(workflowRunTimeout.Round(time.Second).Seconds())
+		taskQueue := &taskqueuepb.TaskQueue{
+			Name: transferTask.TaskQueue,
+			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
 		}
+		workflowRunTimeout := timestamp.DurationValue(executionInfo.WorkflowRunTimeout)
+		taskScheduleToStartTimeoutSeconds := int64(workflowRunTimeout.Round(time.Second).Seconds())
```
Comment (wxing1292) [inline service/history/transferQueueStandbyTaskExecutor.go:174]: this logic will create a mismatch, e.g. sticky task is transmuted into a normal task, but corresponding schedule to start timer still exits

File: service/history/transferQueueActiveTaskExecutor.go [hunk 8]
```diff
@@ -231,11 +232,30 @@ func (t *transferQueueActiveTaskExecutor) processWorkflowTask(
...
+			Name: originalTaskQueue,
+			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+		}
+
+		// Continue to use sticky schedule_to_start timeout as TTL for the matching task. Because the schedule_to_start
+		// timeout timer task is already created which will timeout this task if no worker pick it up in 5s anyway.
+		// There is no need to reset sticky, because if this task is picked by new worker, the new worker will reset
+		// the sticky queue to a new one. However, if worker is completely down, that schedule_to_start timeout task
+		// will re-create a new non-sticky task and reset sticky.
+		err = t.pushWorkflowTask(ctx, task, taskQueue, timestamp.DurationFromSeconds(taskScheduleToStartTimeoutSeconds))
```
Comment (wxing1292) [inline service/history/transferQueueActiveTaskExecutor.go:256]: this special behavior will transmute a sticky workflow task into normal workflow task while the corresponding schedule to start timer can still kill this task.

**Review by wxing1292 — CHANGES_REQUESTED** [review]
I would propose

active logic: fail the sticky workflow task immediately and create a normal workflow task
passive logic: no change

**Comment by wxing1292** [comment]
> I would propose
> 
> active logic: fail the sticky workflow task immediately and create a normal workflow task passive logic: no change

another option is to simply remove sticky task management from history service & make matching service manage the workers (worker with cache or worker without cache)

**Comment by wxing1292** [comment]
> > I would propose
> > active logic: fail the sticky workflow task immediately and create a normal workflow task passive logic: no change
> 
> another option is to simply remove sticky task management from history service & make matching service manage the workers (worker with cache or worker without cache)

yet another solution, add timestamp to respond workflow task completion, do not use sticky queue after e.g. 10m

File: service/history/transferQueueActiveTaskExecutor.go (hunk 4 above)
Comment (yiminc) [reply service/history/transferQueueActiveTaskExecutor.go]: changed to use the sticky ttl to match with the timeout timer.

File: service/history/transferQueueStandbyTaskExecutor.go (hunk 7 above)
Comment (yiminc) [reply service/history/transferQueueStandbyTaskExecutor.go]: that should not matter. when the timer fires, if the task is started, its a no-op; if the task has not started, it just re-create it and clear stickiness, no special logic needed.

File: service/history/transferQueueActiveTaskExecutor.go (hunk 8 above)
Comment (yiminc) [reply service/history/transferQueueActiveTaskExecutor.go:256]: that is true, and it does change the correctness, and it does not degrade the performance.

File: service/history/transferQueueActiveTaskExecutor.go (hunk 4 above)
Comment (meiliang86) [reply service/history/transferQueueActiveTaskExecutor.go]: Retry can take some time, so technically the taskScheduleToStartTimeoutSeconds for the normal task should be the previous value minus the whatever duration is used for the sticky push attempt, is that right?

File: service/history/transferQueueStandbyTaskExecutor.go (hunk 7 above)
Comment (wxing1292) [reply service/history/transferQueueStandbyTaskExecutor.go]: no this PR itself is special handling sticky queue staleness.

anyway, plz put a "experimental" mark here, we need to eventually move the sticky queue management to matching service (source of truth)

**Review by wxing1292 — APPROVED** [review]

File: service/history/transferQueueActiveTaskExecutor.go (hunk 4 above)
Comment (yiminc) [reply service/history/transferQueueActiveTaskExecutor.go]: It does not matter too much. This ttl is used for db to mark the task needs to be deleted so the ttl can be after the actual timeout. There is always the race condition that the task is not TTLed but timeout already fired, so the result is that task will fail to record start and will be discard.

**Pull request merged**